src/main/java/ca/uhn/fhir/jpa/starter/CollectDataResourceProvider.java

```diff
@@ -0,0 +1,155 @@
+package ca.uhn.fhir.jpa.starter;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.servlet.http.HttpServletRequest;
+
+import com.google.common.base.Charsets;
+
+import org.apache.commons.codec.binary.Base64;
+import org.apache.commons.io.IOUtils;
+import org.apache.http.HttpHeaders;
+import org.apache.http.client.ClientProtocolException;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.hl7.fhir.r4.model.Bundle;
+import org.hl7.fhir.r4.model.IdType;
+import org.hl7.fhir.r4.model.Measure;
+import org.hl7.fhir.r4.model.MeasureReport;
+import org.hl7.fhir.r4.model.Observation;
+import org.hl7.fhir.r4.model.Parameters;
+import org.hl7.fhir.r4.model.Reference;
+import org.hl7.fhir.r4.model.Resource;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import ca.uhn.fhir.context.FhirContext;
+import ca.uhn.fhir.jpa.rp.r4.MeasureResourceProvider;
+import ca.uhn.fhir.rest.annotation.IdParam;
+import ca.uhn.fhir.rest.annotation.Operation;
+import ca.uhn.fhir.rest.annotation.OperationParam;
+import ca.uhn.fhir.rest.api.Constants;
+import ca.uhn.fhir.rest.api.EncodingEnum;
+
+public class CollectDataResourceProvider {
+    protected static CloseableHttpClient ourHttpClient;
+
+    @Autowired
+    MeasureResourceProvider measureResourceProvider;
+
+    FhirContext ctx = FhirContext.forR4();
+
+    @Operation(name = PlirConstants.OPERATION_COLLECT_DATA, idempotent = true, type = Measure.class)
+    public Parameters collectDataOperation(HttpServletRequest theServletRequest, @IdParam IdType theId,
+            @OperationParam(name = PlirConstants.PARAM_PERIOD_START, min = 1, max = 1) String periodStart,
+            @OperationParam(name = PlirConstants.PARAM_PERIOD_END, min = 1, max = 1) String periodEnd)
+            throws ClientProtocolException, IOException {
+   
+        Measure measure = this.measureResourceProvider.getDao().read(theId);
+ 
+        if (measure == null) {
+            throw new RuntimeException("Could not find Measure/" + theId.getIdPart());
```

Comment:
Instead of a `RuntimeException`, favour one of the server exceptions [provided by HAPI](https://hapifhir.io/hapi-fhir/apidocs/hapi-fhir-base/ca/uhn/fhir/rest/server/exceptions/package-summary.html). These result in nicer error messages (in this case I think `InvalidRequestException` makes sense). Likewise, `ClientProtocolException` and `IOException` should be wrapped in an `InternalErrorException` and have that thrown instead.

Reply:
@ibacher , i think ResourceNotFoundException is more apropriate in case measure is not found

Reply:
and also   InternalErrorException  could not wrap up   ClientProtocolException and IOException.   I actually even tried BaseServerResponseException and couldnt wrap them also

Reply:
`ResourceNotFound` makes more sense if we're using the `/Measure/[ID]/$collect-data` format.

Reply:
I'm not sure what you mean by the second part. I was thinking this is as simple as:

```java
catch (IOException | ClientProtocolException e) {
	log.error("Caught exception while evaluating measure {}", theId, e);
	throw new InternalErrorException("Could not process measure " + theId, e);
}
```

Reply:
ya i did that , by the way already  [ClientProtocolException](https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/client/ClientProtocolException.html) is a child of  IOException .so we only catch one